lib/seahorn/FiniteMapTransf.cc

```diff
@@ -0,0 +1,338 @@
+  return res;
+}
+
+// ----------------------------------------------------------------------
+//  FiniteMapArgsVisitor
+// ----------------------------------------------------------------------
+
+VisitAction FiniteMapArgsVisitor::operator()(Expr exp) {
+
+  if (isOpX<IMPL>(exp)) { // rule (or implication inside rule?)
+    Expr head = exp->right();
+    Expr body = exp->left();
+    assert(body); // facts!
+    Expr fdecl = *head->args_begin();
+
+    if (m_pred_decl_t.count(fdecl) == 0)
+      return VisitAction::changeDoKids(exp);
+
+    Expr newPredDecl = m_pred_decl_t.find(fdecl)->second;
+    ExprVector newUnifs;
+    Expr newFapp = mkFappArgsCore(head, newPredDecl, m_evars, m_efac, newUnifs);
+    Expr newBody = newUnifs.empty() ? body : mk<AND>(mknary<AND>(newUnifs), body);
+
+    Expr newExp = boolop::limp(newBody, newFapp);
+    // efficiency: are we traversing the newly created unifs?
+    return VisitAction::changeDoKids(newExp);
+
+  } else if (isOpX<FAPP>(exp) &&
+             !bind::IsConst()(exp)) { // faster to check arity >= 2?
+    Expr fdecl = *exp->args_begin();
+    if (m_pred_decl_t.count(fdecl) > 0) { // needs to be transformed
+      ExprVector newUnifs;
+      Expr newPredDecl = m_pred_decl_t.find(fdecl)->second;
+      Expr newExp = mkFappArgsCore(exp, newPredDecl, m_evars, m_efac, newUnifs);
+      if (!newUnifs.empty())
+        newExp = mk<AND>(mknary<AND>(newUnifs), newExp);
+      LOG("fmap_transf", errs() << *newExp << "\n";);
+      return VisitAction::changeDoKids(newExp);
+    }
+  } else if (isOpX<FDECL>(exp)) {
+    return VisitAction::skipKids();
+  }
+  return VisitAction::doKids();
+}
+
+// ----------------------------------------------------------------------
+//  FiniteMapBodyVisitor
+// ----------------------------------------------------------------------
+
+static bool returnsFiniteMap(Expr e) {
+  return isOpX<CONST_FINITE_MAP>(e) || isOpX<SET>(e) ||
+         bind::isFiniteMapConst(e);
+}
+
+VisitAction FiniteMapBodyVisitor::operator()(Expr exp) {
+  LOG("fmap_transf", errs() << "Creating visit action for: " << *exp << "\n";);
+
+  if (isVisitFiniteMapOp(exp)) {
+    return VisitAction::changeDoKidsRewrite(exp, m_rw);
+    } else if (bind::isFiniteMapConst(exp)) {
+      return VisitAction::changeDoKidsRewrite(exp, m_rw);
+    } else if (isOpX<EQ>(exp)) {
+      if (returnsFiniteMap(exp->left()) || returnsFiniteMap(exp->right()))
+        return VisitAction::changeDoKidsRewrite(exp, m_rw);
+  } else if (bind::IsConst()(exp) || bind::isFdecl(exp)) {
+      return VisitAction::skipKids();
+    }
+    // The step doesn't need to be rewritten but the kids do
+    return VisitAction::doKids();
+  }
+
+bool FiniteMapBodyVisitor::isVisitFiniteMapOp(Expr e) {
+    return isOpX<CONST_FINITE_MAP>(e) || isOpX<GET>(e) || isOpX<SET>(e);
+    // we are not visiting CONST_FINITE_MAP_KEYS, DEFAULT,
+  }
+
+} // namespace seahorn
+
```

Comment:
this was too much for me to review. Consider using classes to group related methods together.  Right now, it is very hard to understand how the state (i.e., various maps) are passed between functions and whether this is done correctly.

This comment can be left for refactoring after the PR is merged.

Reply:
I created a struct `FMapExprsInfo` to store all objects used for caching during rewritting. I also grouped the rewritting functions by visitor and added documentation.